### ISSUE TITLE
`buildSrc` kotlin compilation in config phase doesn't use `kotlin.daemon.jvmargs` from root `gradle.properties`, duplicating the KotlinCompileDaemon

### ISSUE THREAD

**Issue by rolgalan** [gradle/gradle#25960]
### Expected Behavior

KotlinCompileDaemon should use same settings during scripts compilation and during project compilation, allowing to reuse it in both parts of the builds.

### Current Behavior

KotlinCompileDaemon uses different memory settings to be initialized during scripts compilation and during project compilation, so two daemons are spawn during the build.

### Context (optional)

I was inspecting a build with a visual memory profiler and I noticed the `KotlinCompileDaemon` got duplicated in the middle of the execution, when the modules started to execute their `compileKotlin` tasks.

![Screenshot 2023-07-31 at 11 33 36](https://github.com/gradle/gradle/assets/14839961/c50891ca-a78d-405d-aac3-259d2ebc6bb5)

There was a difference in JVM arguments that made Gradle spawn a new daemon as was unable to reuse the existing one due a different value of `Xmx`: `-Xmx7g` vs `-Xmx4g`.

I have this in my gradle.properties file:

```
org.gradle.jvmargs=-Xmx7g
kotlin.daemon.jvmargs=-Xmx4g
```

The initial daemon created during the configuration phase for the kotlin scripts was using `Xmx7g`, so it was not using the specific daemon configuration and inheriting from the main Gradle Daemon. The second daemon, created to compile the project code, was properly using the `Xmx4g` as expected.

Tested this with Gradle `8.3-rc2`, `8.2.1`, both with same behaviour.
I catch this locally (with macOs), but I found it also in our CI on linux machines.

Since we are using the [InfoKotlinProcess](https://github.com/cdsap/InfoKotlinProcess) plugin, the duplicated daemon can be also seen in the scans:

![Screenshot 2023-07-31 at 12 19 46](https://github.com/gradle/gradle/assets/14839961/6177cbb5-1028-4ed7-ae16-672fdc58960c)

### Steps to Reproduce

Set different memory settings for Gradle and Kotlin Daemons, for example:
```
org.gradle.jvmargs=-Xmx7g
kotlin.daemon.jvmargs=-Xmx4g
```

Open a project with both kotlin `.kts` gradle script files and kotlin source code. In this case, the project uses a `buildSrc` folder, didn't try without it.

Build the project and watch the local jvm processes.

### Gradle version

8.3-rc2, 8.2.1 

### Your Environment (optional)

OpenJDK Runtime Environment Temurin-17.0.6+10 (build 17.0.6+10)
Both Linux and macOS.

**Comment by eskatos** [comment]
Sorry that you're having trouble with Gradle!

We appreciate the effort that went into filing this issue, but we must ask for more information. 

As stated in our issue template, a [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example) is a must for us to be able to track down and fix your problem efficiently. Our available resources are severely limited, and we must be sure we are looking at the exact problem you are facing.

If we have a reproducer, we may be able also to suggest workarounds or ways to avoid the problem.

The ideal way to provide a reproducer is to leverage our [reproducer template](https://github.com/gradle/gradle-issue-reproducer). You can also use [Gradle Project Replicator](https://github.com/android/project-replicator) to reproduce the structure of your project. 

This issue will be closed after 7 days unless you provide more information.

----

Gradle script compilation happens in the Gradle daemon, not in a Kotlin compiler daemon.

I was not able to reproduce with: https://github.com/eskatos/issue-reproducers/tree/main/gradle/issue-25960

Please provide a valid reproducer.

**Comment by rolgalan** [comment]
> This issue will be closed after 7 days unless you provide more information.
> 
> Gradle script compilation happens in the Gradle daemon, not in a Kotlin compiler daemon.
> 
> I was not able to reproduce with: https://github.com/eskatos/issue-reproducers/tree/main/gradle/issue-25960
> 
> Please provide a valid reproducer.

Hi @eskatos thanks for providing a reproducer yourself! ❤️ 

I've been able to duplicate the Dameon with your reproducer, but taking a different route. In my project it happened directly with every clean build, but we have many things in the project, so not sure what could be causing this. In any case, this shows that the behaviour is inconsistent. Let me share how to reproduce the issue :)

If I run `./gradlew build`, I see a KotlinCompileDaemon being spawn when the `:buildSrc:compileKotlin` starts. This daemon has `-Xmx7g`, therefore is not using the `kotlin.daemon.jvmargs`. Later the `:compileKotlin` task somehow reuses this existing Daemon. So no duplication during this execution, but **the memory settings are incorrect**.

If I run now `./gradlew clean build`, so the buildSrc remains there, for the `:compileKotlin` task it keeps reusing the existing 7g KotlinDaemon even if it has different memory setting. Not duplication yet.

If you kill that 7g KotlinDaemon and execute again `./gradlew clean build`, the new KotlinDaemon spawn here will have `-Xmx4g`, therefore this one honors the `kotlin.daemon.jvmargs`. Since we killed the daemon, this is not a duplication, but **this proves that the buildSrc and project compileKotlin tasks use different sources/settings to create the kotlin daemon**.

Now, leaving that Daemon there, if we force to rebuild the buildSrc again (`rm -rf buildSrc/build`), when executing `./gradlew build` it ignores the existing 4g KotlinDameon and spawns a new one with `-Xmx7g`. All successive builds will keep using this new 7g daemon. While the other 4g daemon is there, ending up with 2 duplicated KotlinDaemons in your system.

This looks hard to replicate. But a developer might end up in this state simply if they start a session with their scripts already compiled, and introduce a change in them, forcing them to rebuild.

So, to summarize:
- `./gradlew build`  <-- creates 7g daemon
- `kill ${daemon_pid}`  <-- kill 7g daemon
- `./gradlew clean build` <-- creates 4g daemon
- `rm -rf buildSrc/build`
- `./gradlew clean build`<-- creates new 7g daemon ignoring the existing

Also we just verified that, by copying the `gradle.properties` file inside the `buildSrc` folder, it reads properly the jvmargs. If we repeat above steps now, it successfully uses a single 4g daemon all the time.

Let me know if this helps or do you need anything else from my side :)

**Comment by eskatos** [comment]
Thank you for the detail, much appreciated.

I rephrased the title of the issue to better reflect that this is about `buildSrc` Kotlin compilation, not about build script compilation.

The fact that if a 7g daemon is available when the 4g one is requested then the 7g is reused is because the Kotlin compiler infrastructure assumes that the existing 7g daemon is "compatible" with the 4g requirement.

Having the `buildSrc/gradle.properties` file is the way to go to have a single Kotlin compiler daemon. It will be the same for any included build.

All this makes sense and is working as intended. What you could do is raise this with the Kotlin team so that their documentation include guidance on how to configure the Kotlin compiler daemon in presence of `buildSrc` or included builds.

I'm closing this issue as from Gradle pov it works as intended.

**Comment by rolgalan** [comment]
Thanks for the clarification @eskatos. I was not aware that it was intended to use a separate `gradle.properties`, I assumed that it would inherit the parent configuration in the absence of any. It's clear now! 👍
